README.md

```diff
@@ -1 +1,76 @@
+
+// Now `token` has all the PSP22 methods
+let balance = token.balance_of(some_account);
+token.transfer(recipient, value); // returns Result<(), PSP22Error>
+```
+
+The same method can be used with other traits (`PSP22Metadata`, `PSP22Burnable`, `PSP22Mintable`) defined in this crate. See the contents of [`traits.rs`][traits].
+
+### 2. Ready to use contract
+
+The file [`lib.rs`][lib] contains a ready to use implementation of basic PSP22 token contract (extended with PSP22Metadata). To use it, please check out this repository and compile its contents with [`cargo-contract`][cargo-contract] with the `"contract"` feature enabled:
+```
+$ cargo contract build --release --features "contract"
+```
+
+### 3. Custom implementation of PSP22 logic with `PSP22Data`
+
+The `PSP22Data` class can be used to extend your contract with PSP22 token logic. In other words, you can easily build contracts that implement PSP22 interface alongside some other functionalities defined by the business logic of your project.
+
+The methods of the `PSP22Data` class correspond directly correspond to queries and operations defined by the PSP22 token standard. To make your contract become a PSP22 token, you need to:
+ - Put a single `PSP22Data` instance in your contract's storage and initialize it with some starting supply of tokens.
+ - Add definitions of `Transfer` and `Approval` events in the body of your contract.
+ - Add the `impl PSP22 for [struct_name]` block with implementation of PSP22 trait messages using `PSP22Data` methods. Each method which mutates the state of the token database returns a `Result<Vec<PSP22Event>, PSP22Error>` with all events generated by that operation. Please make sure to handle errors correctly and emit the resulting events (see the `emit_events` function).
+ - Optionally implement also the `PSP22Metadata` trait to make your token play nice with other ecosystem tools.
+
+The contract in [`lib.rs`][lib] contains an example implementation following all the above steps. Feel free to copy-paste parts of it.
+
+### 4. Burnable and Mintable extensions
+
+The `PSP22Data` class contains also `burn` and `mint` methods, which can be used to implement `PSP22Burnable` and `PSP22Mintable` extensions and make your token burnable and/or mintable. An example implementation follows the same pattern as for the base trait:
+```
+impl PSP22Burnable for Token {
+    #[ink(message)]
+    fn burn(&mut self, value: u128) -> Result<(), PSP22Error> {
+        // Check if the caller is allowed to burn!
+        let events = self.data.burn(self.env().caller(), value)?;
+        self.emit_events(events);
+        Ok(())
+    }
+}
+```
+Please note that `PSP22Data` `burn` and `mint` methods do not enforce any form of access control. It's probably not a good idea to have a token which can be minted and burned by anyone anytime. When implementing Burnable and Mintable extensions, please make sure that their usage is restricted according to your project's business logic.
```

Comment:
I'd consider providing an example implementation of restricting these to only owner

Reply:
That's maybe a bit outside of the scope of this readme, but ok, it won't hurt ;)